File: src/leetCodeProblems/BinarySearch/CountPairsWithGivenDifference2006.java
package leetCodeProblems.BinarySearch;

/**
 * Leet Code - https://leetcode.com/problems/count-number-of-pairs-with-absolute-difference-k/
 */

import java.util.*;

public class CountPairsWithGivenDifference2006 {

	private int binarySearch(int[] arr, int X, int low)  {
        int high = arr.length - 1;
         
        int ans = arr.length;
         
        while(low <= high) {
            int mid = low + (high - low) / 2;
            if(arr[mid] >= X) {
                ans = mid;
                high = mid - 1;
            }
            else low = mid + 1;
        }
        return ans;
    }
    
    
    public int countKDifference(int[] nums, int k) {
        
        Arrays.sort(nums);
        
        int count = 0;
        
        for(int i=0; i<nums.length; i++) {

            int targetNumUsingSum = k+nums[i];
            
            //System.out.println("nums[i] ->" + nums[i]);
            
            int searchIndex = binarySearch(nums, targetNumUsingSum, i+1);
            
            if (searchIndex != nums.length) {
                int incrementalSearchIndex = binarySearch(nums, targetNumUsingSum+1, i+1);
                count += incrementalSearchIndex - searchIndex;
            }
            
            //System.out.println("searchIndex ->" + searchIndex);
            
        }
        
        return count;
        
    }
    
    public static void main(String[] args) {
				
		int[] inputArray = {1, 2, 2, 1};
		int targetDifference = 1;
		
		CountPairsWithGivenDifference2006 obj = new CountPairsWithGivenDifference2006();
		
		System.out.println(obj.countKDifference(inputArray, targetDifference));
	}
}

File: src/leetCodeProblems/BinarySearch/CountPairsWithGivenDifference2006.go
package main

import (
	"log"
	"sort"
)

/*
- LeetCode - https://leetcode.com/problems/count-number-of-pairs-with-absolute-difference-k/description/
- Time - O(nlogn)
- Space - O(1)
*/

func binarySearchWithDuplicatesHandling(nums []int, start, search int) (int, bool) {

	end := len(nums) - 1
	out := len(nums)

	for start <= end {
		mid := start + (end-start)/2

		if nums[mid] >= search {
			out = mid
			end = mid - 1
		} else {
			start = mid + 1
		}
	}

	return out, false
}

func countKDifference(nums []int, k int) int {
	sort.Ints(nums)
	out := 0

	for i, v := range nums {
		if searchIndex, _ := binarySearchWithDuplicatesHandling(nums, i+1, v+k); searchIndex != len(nums) {
			incrementedSearchIndex, _ := binarySearchWithDuplicatesHandling(nums, i+1, v+k+1)
			out += incrementedSearchIndex - searchIndex
		}
	}

	return out
}

func main() {

	//log.Println(countKDifference([]int{1, 2, 2, 1}, 1))
	log.Println(countKDifference([]int{10, 2, 10, 9, 1, 6, 8, 9, 2, 8}, 5))
}
